File: Editor/Tools/GetEditorStateTool.cs
using UnityEditor;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Lightweight query tool for Unity Editor state flags (Playing/Compiling/etc).
    /// Side-effect free; safe to call during Play Mode without risking Auto Refresh freeze.
    /// Agents should precheck this before any file-change / assembly-reload inducing call.
    /// </summary>
    public class GetEditorStateTool : McpToolBase
    {
        public GetEditorStateTool()
        {
            Name = "get_editor_state";
            Description = "Returns Unity Editor state flags (isPlaying, isPlayingOrWillChangePlaymode, " +
                          "isCompiling, isUpdating, isPaused). Safe to call anytime — no side effects, " +
                          "no Asset Pipeline triggers. Use this before execute_csharp / recompile_scripts " +
                          "to avoid Play Mode freeze.";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = "Editor state snapshot",
                ["isPlaying"] = EditorApplication.isPlaying,
                ["isPlayingOrWillChangePlaymode"] = EditorApplication.isPlayingOrWillChangePlaymode,
                ["isCompiling"] = EditorApplication.isCompiling,
                ["isUpdating"] = EditorApplication.isUpdating,
                ["isPaused"] = EditorApplication.isPaused
            };
        }
    }
}

File: Server~/src/tools/getEditorStateTool.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import * as z from 'zod';
import { Logger } from '../utils/logger.js';

const toolName = 'get_editor_state';
const toolDescription =
  'Returns Unity Editor state flags (isPlaying, isPlayingOrWillChangePlaymode, isCompiling, ' +
  'isUpdating, isPaused). Side-effect free — safe to call anytime without triggering Asset ' +
  'Pipeline Refresh or assembly reload. Use this before execute_csharp / recompile_scripts / ' +
  'any file-mutating call to avoid Play Mode freeze.';

const paramsSchema = z.object({});

export function registerGetEditorStateTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(mcpUnity: McpUnity, params: any) {
  const response = await mcpUnity.sendRequest({
    method: toolName,
    params
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to fetch editor state'
    );
  }

  const flags = {
    isPlaying: !!response.isPlaying,
    isPlayingOrWillChangePlaymode: !!response.isPlayingOrWillChangePlaymode,
    isCompiling: !!response.isCompiling,
    isUpdating: !!response.isUpdating,
    isPaused: !!response.isPaused
  };

  const summary = Object.entries(flags)
    .map(([k, v]) => `${k}=${v}`)
    .join(', ');

  return {
    content: [{
      type: 'text' as const,
      text: `Editor state: ${summary}`
    }],
    data: flags
  };
}
